### DISCUSSION TITLE
How does token.immediate interact with repeat and seq?

### DISCUSSION THREAD

**Discussion by ahelwer** [tree-sitter/tree-sitter#1249] [Q&A] [answered]
Something I've run into a couple times and never figured out. Say we have some rules like:
```
nonterminal: $ => seq(
  'X',
  token.immediate($.terminal),
  repeat(seq($.nonterminal, $.terminal)),
  token.immediate('X')
),
terminal: $ => 'Y'
```
How do I interleave `token.immediate` with `repeat` and `seq` to ensure no whitespace/extras are allowed in?

**Comment by ahlinc** [comment, marked as answer]
I also tried to solve similar tasks, but currently there are issues with it #1156 and it's [impossible for now](https://github.com/tree-sitter/tree-sitter/issues/449#issuecomment-533702316) to use other rule in the `token()` and in the `token.immediate()`. Also there was suggestion to simplify this in some way #1102.

You came up with a rather complex grammar example straight away :smile: It has recursion and is a palindrome.
As a workaround with `token.immediate` I found that it's pretty useful to write [constructor functions](https://github.com/ahlinc/tree-sitter-plantuml/blob/0cc17078ae170801e93c44a2f9e80eef81cad2f7/grammar.js#L458-L557) and [use these](https://github.com/ahlinc/tree-sitter-plantuml/blob/0cc17078ae170801e93c44a2f9e80eef81cad2f7/grammar.js#L205-L212) in a grammar rules.

_For your example grammar:_
```js
module.exports = grammar({
  name: "d1249",

  rules: {
    nonterminal: $ => nonterminal(alias($.nonterminal_inner, $.nonterminal)),
    nonterminal_inner: $ => prec.left(nonterminal(alias($.nonterminal_inner, $.nonterminal), i)),
  },
});

function nonterminal(ref, w = x => x) {
  return seq(
    w('X'),
    alias(i('Y'), sym('terminal')),
    repeat(seq(ref, alias(i('Y'), sym('terminal')))),
    i('X')
  )
}

// Shorter wrapper on token.immediate()
// May be called as i('token_string') or as i`token_string` for simple cases.
function i(s) {
  if (Array.isArray(s) && s.hasOwnProperty('raw'))
    return token.immediate(s[0]);
  else return token.immediate(s);
}

/* Can be checked with:
   # tree-sitter p <(echo -en "XYXYXYXYXYXYXYX")
*/
```
![Screenshot from 2021-07-10 03-55-35](https://user-images.githubusercontent.com/14666676/125147239-bbc81b00-e132-11eb-9499-66f30caac8ac.png)

Another fun situation with the above palindrome is that it starts and ends with the same `'X'` symbol so the parser prefers to do the `repeat` step instead of making recursion and if to replace the last `'X'` to `'Z'` then it will make recursion.
```js
function nonterminal(ref, w = x => x) {
  const terminal = alias(i('Y'), sym('terminal'));
  return seq(
    w('X'),
    terminal,
    repeat(seq(ref, terminal)),
    i('Z') // 'X' was replaced to 'Z'
  )
}

/* Can be checked with:
   # tree-sitter p <(echo -en "XYXYXYXYZYZYZYZ")
*/
```
![Screenshot from 2021-07-10 04-03-13](https://user-images.githubusercontent.com/14666676/125147436-f41c2900-e133-11eb-95db-7edebbd3619f.png)

**As conclusion with current tree-sitter state related to `token.immediate()`:**
- If unnamed nodes required for analysis than It can be applied only to simple strings, with regexes and other cases described in #1156 there are unnamed nodes loss _(I starting suspect that it's intentional behavior for some cases from #1156)._
  - It's possible to use other DSL functions inside of `token()` and `token.immediate()` and by default these assume that there would be no spaces between any inner tokens, so the `token.immediate()` comparing to the `token()` is only about to check that there are no spaces before the token described by it.
  - If the absence of unnamed nodes isn't the problem than any other DSL functions **but not rule references** can be used inside of `token()` and `token.immediate()` functions.
    - The `alias(i('Y'), sym('terminal')` construct is the workaround for impossibility to use rule references inside of `token()` and `token.immediate()` functions.
- Can be used inside of `seq()`, `repeat()/repeat1()` and `choice()` normally.
- If used as the first token in a rule than such rule start is an `immediate` and requires to have no spaces before it.
  - To handle this dynamically the `w()` wrapper on the first rule symbol is used on above examples in joint with `i()` wrapper passed to the constructor function where it's needed to produce the `immediate` rule definition.
- The `sym('rule_name')` used as equivalent for `$.rule_name` to don't pass `$` additionally to every function that require to use rule references.
